Enable logexporter mechanism to dump logs from k8s nodes to GCS directly

shyamjvs opened the pull request:
Ref https://github.com/kubernetes/kubernetes/issues/48513

This adds support for logexporter from k8s side. Next I'll send a PR adding support from test-infra side.

/cc @kubernetes/sig-scalability-misc @kubernetes/test-infra-maintainers @fejta @wojtek-t @gmarek

shyamjvs committed:
Enable logexporter mechanism to dump logs from k8s nodes to GCS directly

wojtek-t commented on cluster/log-dump/log-dump.sh:
```diff
@@ -0,0 +1,316 @@
+  # Sleep for logexporter_sleep_seconds before killing the daemonset. There are 2 cases here:
+  # 1. A daemon instance finished log upload already -
+  #    This is fine as it would be sleeping by then as the sleep cannot finish before we call the delete below.
+  #    (It can happen in a highly unlikely case of the instance finishing upload before the create command above returns.)
```
I don't understand it.

shyamjvs replied:
Assume the following series of events:
(create daemonset request sent) -> (daemonset created) -> (pod created) -> (pod uploaded logs and sleep started) -> (response for create daemonset request received (a bit late due to some delay in apiserver/network)) -> (sleep before 'delete daemonset' called, but sleep for pod already started a while ago so pod finishes first before deleting the daemonset)

wojtek-t commented on cluster/log-dump/log-dump.sh:
```diff
@@ -0,0 +1,316 @@
+  sed -i'' -e "s/{{.CloudProvider}}/${cloud_provider}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+  sed -i'' -e "s/{{.GCSPath}}/${gcs_artifacts_dir}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+  sed -i'' -e "s/{{.EnableHollowNodeLogs}}/${enable_hollow_node_logs}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
+  sed -i'' -e "s/{{.SleepSeconds}}/${logexporter_sleep_seconds}/g" "${KUBE_ROOT}/cluster/log-dump/logexporter-daemonset.yaml"
```
I would do sth like 2*logexporter_sleep_second to avoid races etc.

wojtek-t replied:
Or actually sth that I suggested at some point - replace with something supper huge to really guarantee that it won't finish.

shyamjvs replied:
Could you clarify race with what? I think the only race is b/w the pod dying after finishing timeout and daemonset deleting. If the latter happens first, it's fine. And the former IIUC can't happen first because time taken by it would be (pod creation + pod uploading logs + sleep) whereas time taken by latter is just (sleep).

wojtek-t commented on cluster/log-dump/log-dump.sh:
```diff
@@ -0,0 +1,316 @@
+
+  # Sleep for logexporter_sleep_seconds before killing the daemonset. There are 2 cases here:
+  # 1. A daemon instance finished log upload already -
+  #    This is fine as it would be sleeping by then as the sleep cannot finish before we call the delete below.
```
if you proceed with infinite timeout, the you don't need this sentence.
Actually, I would get rid of "timeout" parameter in yaml template and just set it to sth super huge.

shyamjvs replied:
Sounds reasonable. However even if we put a huge timeout, we still need a timeout for deleting the daemonset as we can't keep waiting indefinitely for the logexporting step. In that case, we might as well add a timeout to the pod (that is slightly greater than the timeout for the logexporting step).

wojtek-t replied:
Yes - we still need this timeout. But less fields we make configurable in the manifest, less errors will happen :)

shyamjvs replied:
SGTM. We can get rid of it since we're going to check the status now.

wojtek-t commented on cluster/log-dump/log-dump.sh:
```diff
@@ -0,0 +1,316 @@
+  #    TODO(shyamjvs): Call delete earlier if we know all daemonsets finished, to save some time.
+  # 2. A daemon instance has not uploaded the logs yet -
+  #    We still have to delete the daemonset to respect timeout. And logexporter_sleep_seconds has been chosen liberally
+  #    enough to make this unlikely. We can increase it in future if needed (as logexporter is already saving much time).
```
Now when I think about it, I think we need to be able to say whether the jobs finished or not.
Currently, if there is no log file for a given node you don't know if:
- log exporter didn't upload it, because of lack of time/error/...
- there wasn't such file on the node at all
- something else?

When I know think about it, I think we really need to expose sth like "/status" endpoint in the logexporter job and to excercise it for all nodes to check how many of them finished and what nodes didn't finish...

shyamjvs replied:
Adding a /status endpoint is a good idea. @gmarek was also suggesting something similar. But IMO it would be an overkill as IIUC then you need to hit 5000 nodes' /status endpoints periodically to check which all finished. If you just want to know which nodes uploaded logs we can rather just do a `gsutil ls` on the artifacts dir. And if some node's logs are missing:

> log exporter didn't upload it, because of lack of time/error/...

We know the node, so we can try the traditional method for the node (through gcloud ssh).

> there wasn't such file on the node at all

This will not lead to logexporter failing. It will just skip that file (similar to what we're currently doing). And logexporter will actually upload the logs it managed to find as you'd expect.

Does that sound reasonable?

wojtek-t replied:
>> there wasn't such file on the node at all

> This will not lead to logexporter failing. It will just skip that file (similar to what we're currently doing). And logexporter will actually upload the logs it managed to find as you'd expect.

But how the end user then know if this is log-exporting issue or lack of file. If the exporting script will explicitly write "all jobs finished successfully" then he knows.
With the design you're introducing the user is not able to distinguish a situation of: log exporter didn't do the job vs no files to export by it. We need to be able to distinguish it.

wojtek-t replied:
And I'm fine with not hiting /status endpoint periodically as a starting point. Just doing it at the end to check if the logs were grabbed or not is enough.

shyamjvs replied:
SGTM.

> But how the end user then know if this is log-exporting issue or lack of file.

He can do a `kubectl logs` on the pod. In fact we can just do kubectl logs instead of status endpoint and check for the line "Logs successfully uploaded". If it is present it means logexporter succeeded and if some file is missing just means it wasn't there at all. Otherwise we just assume the pod failed (it could still be uploading logs but we just assume it failed as we're checking only once). We can also print logs from each of those pods for end-user info. SGTY?

wojtek-t commented:
@gmarek

shyamjvs commented:
@wojtek-t Addressed review comments solving the timeout problem by probing pod logs. Further I changed it to do regular log-dump (through SSH) for nodes failing logexporter, making it quite safe to use.

gmarek commented on cluster/log-dump/logexporter-daemonset.yaml:
```diff
@@ -0,0 +1,73 @@
+    spec:
+      containers:
+      - name: logexporter-test
+        image: gcr.io/google-containers/logexporter:latest
```
Please don't do latest.

shyamjvs replied:
Done.

wojtek-t commented on cluster/log-dump/log-dump.sh:
```diff
@@ -0,0 +1,344 @@
+  for node_name in "${node_names[@]}"; do
+    echo -n "Checking logexporter status for node ${node_name}..."
+
+    pod_name=$(kubectl describe node ${node_name} 2>&1 | grep -o "logexporter-[0-9a-z]*" -m 1) || true
```
Calling kubectl has a significant overhead, because it resolves a bunch of things at the beginning. I'm afraid that if you're calling kubectl twice for each node, this will result it being able to process ~1pod/s.

That said:
1. to get pod_name, you can simply list all pods from the logexporter namespace and the simply check for each node if there is a pod in this list for that node.
2. I'm not sure with "kubectl get logs" will be fast enough, but we can start with this. But please at least add a TODO.

gmarek replied:
+1 it's better to do `kubectl get pod -o wide | grep node_name` to get a pod name (or use jsonpath format to be even nicer)

shyamjvs replied:
Done. It wasn't as easy as I expected (had to experiment multiple times before making bash work).

wojtek-t commented on cluster/log-dump/logexporter-daemonset.yaml:
```diff
@@ -0,0 +1,73 @@
+        - --gcs-path={{.GCSPath}}
+        - --gcloud-auth-file-path=/etc/service-account/service-account.json
+        - --enable-hollow-node-logs={{.EnableHollowNodeLogs}}
+        - --sleep-duration=1h
```
Make it even bigger. E.g. 24h or sth like that.

shyamjvs replied:
Done.

gmarek commented on cluster/log-dump/log-dump.sh:
```diff
@@ -0,0 +1,344 @@
+#!/bin/bash
```
This is a move with some changes, right? Are you sure you deleted old 'log-dump.sh', or are you leaving it there for a reason?

shyamjvs replied:
Not deleting the old log-dump.sh just yet because then the current kubetest (which shells out to the script) will fail. After this PR is pushed, I'll reflect the change in test-infra and push a new image. Then we can delete it.

gmarek commented on cluster/log-dump/log-dump.sh:
```diff
@@ -0,0 +1,344 @@
+function dump_nodes_with_logexporter() {
+  echo "Detecting nodes in the cluster"
+  detect-node-names &> /dev/null
+  node_names=( "${NODE_NAMES[@]}" )
```
unused

shyamjvs replied:
Removed. Thanks.

gmarek commented on cluster/log-dump/log-dump.sh:
```diff
@@ -0,0 +1,344 @@
+      continue
+    fi
+
+    pod_success_log=$(kubectl get logs ${pod_name} -n logexporter 2>&1 | grep "Logs successfully uploaded") || true
```
I still think that having an http endpoint on all those daemons would be better than depending on logs, but as Wojtek wrote - it's fine for now. Please add a TODO.

shyamjvs replied:
Added a TODO. But I'm not sure how having a /status endpoint helps. It is still O(#nodes) work. It probably eliminates doing 'kubectl logs' (btw logs are really small for logexporter pods < 1kB)  but instead needs something like 'kubectl proxy' iiuc. What is more important is to be able to parallelize these kubectl calls IMO and let's see if we actually need that in practice. Added a TODO for this too.

shyamjvs replied:
cc @wojtek-t

shyamjvs replied:
I meant 'kubectl exec' above instead of proxy - sorry.

gmarek replied:
Ugh, right - you can't easily distinguish Deamons and create services for them, which would solve this problem:/

shyamjvs replied:
Yup. We need to hit specific endpoints instead of a load-balanced service ip.

fejta commented:
/test pull-kubernetes-e2e-kops-aws

shyamjvs commented:
@wojtek-t @gmarek Addressed the comments. PTAL now.

wojtek-t commented on cluster/log-dump/log-dump.sh:
```diff
@@ -0,0 +1,363 @@
+  # TODO(shyamjvs): Parallelize the for loop below to make it faster (if needed).
+  logexporter_pods=( $(echo "${pods_and_nodes}" | awk '{print $1}') )
+  logexporter_nodes=( $(echo "${pods_and_nodes}" | awk '{print $2}') )
+  for index in "${!logexporter_pods[@]}"; do
```
What is this "!" here?

shyamjvs replied:
It is a convenient way of iterating through 0-(length-1). Otherwise the expression would look ugly as we have to calculate length-1.

wojtek-t replied:
OK - bash will always be surprising me.

shyamjvs replied:
Always!

wojtek-t reviewed (commented):
Just single question - other than that LGTM.

wojtek-t commented:
/lgtm

shyamjvs commented:
/test pull-kubernetes-federation-e2e-gce

gmarek commented:
/retest
